compel/test/stack/spy.c

```diff
@@ -0,0 +1,368 @@
+		ret = -1;
+		goto done;
+	}
+
+	char victim_maps_path[6 + 11 + 5 + 1];
+	int written = snprintf(victim_maps_path, sizeof(victim_maps_path), "/proc/%d/maps", pid);
+	if (written < 0 || written >= sizeof(victim_maps_path)) {
+		fprintf(stderr, "Failed to create path string to victim's /proc/%d/maps file\n", pid);
+		ret = -1;
+		goto done;
+	}
+
+	maps = fopen(victim_maps_path, "r");
+	if (maps == NULL) {
+		perror("Can't open victim's /proc/$pid/maps");
+		ret = -1;
+		goto done;
+	}
+
+	while (getline(&maps_line, &maps_line_len, maps) != -1) {
+		unsigned long vmstart, vmend;
+		if (sscanf(maps_line, "%lx-%lx %c%c", &vmstart, &vmend, &r, &w) < 4) {
+			fprintf(stderr, "Can't parse victim's /proc/%d/maps; line: %s\n", pid, maps_line);
+			ret = -1;
+			goto free_linebuf;
+		}
+
+		if (page >= (void *)vmstart && page < (void *)vmend) {
+			if (w == 'w') {
+				if (r != 'r') {
+					fprintf(stderr, "Expecting writable memory to also be readable");
+					ret = -1;
+					goto free_linebuf;
+				}
+				ret = 1;
+			}
+			break;
+		}
+	}
+	if (errno) {
+		perror("Can't read victim's /proc/$pid/maps");
+		ret = -1;
+	}
+free_linebuf:
+	free(maps_line);
+	fclose(maps);
+done:
+	return ret;
+}
+
+static void *read_proc_mem(int pid, void *offset, size_t len)
+{
+	char victim_mem_path[6 + 11 + 4 + 1];
+	int written = snprintf(victim_mem_path, sizeof(victim_mem_path), "/proc/%d/mem", pid);
+	if (written < 0 || written >= sizeof(victim_mem_path)) {
+		fprintf(stderr, "Failed to create path string to victim's /proc/%d/mem file\n", pid);
+		return NULL;
+	}
+	int fd = open(victim_mem_path, O_RDONLY);
+	if (fd < 0) {
+		perror("Failed to open victim's /proc/$pid/mem file");
+		return NULL;
+	}
+	void *data = malloc(len);
+	ssize_t read = pread(fd, data, len, (off_t)offset);
+	if (read == -1) {
+		perror("Failed to read victim's /proc/$pid/mem file");
+		goto freebuf;
+	}
+
+	return data;
+
+freebuf:
+	free(data);
+	return NULL;
+}
+
+static int save_data_near_stack(struct parasite_ctl *ctl, int pid, void *stack, void **saved_data,
+				size_t *saved_data_size)
+{
+	size_t page_mask = page_size() - 1;
+	size_t saved_size = 0;
+	size_t stack_size_last_page = (uintptr_t)stack & page_mask;
+	void *next_page = stack;
+	if (stack_size_last_page != 0) {
+		size_t empty_space_last_page = page_size() - stack_size_last_page;
+		saved_size = min(empty_space_last_page, (size_t)SAVED_DATA_MAX);
+		next_page += page_size() - stack_size_last_page;
+	}
+	while (saved_size < SAVED_DATA_MAX && next_page != NULL) {
+		switch (page_writable(ctl, pid, next_page)) {
+		case 1:
+			saved_size = min((size_t)(saved_size + page_size()), (size_t)SAVED_DATA_MAX);
+			next_page += page_size();
+			break;
+		case 0:
+			next_page = NULL;
+			break;
+		default:
+			return -1;
+		}
+	}
+	if (saved_size > 0) {
+		void *sd = read_proc_mem(pid, stack, saved_size);
+		if (sd == NULL)
+			return -1;
+		*saved_data = sd;
+	} else
+		*saved_data = NULL;
```

Comment:
All branches of the if else statement should either have brackets simultaneously or don't have them simultaneously. The above is not allowed by codding style. https://www.kernel.org/doc/html/v4.10/process/coding-style.html#placing-braces-and-spaces

Reply:
Fixed, thanks.